### DISCUSSION TITLE
How to use a single Backspace press to delete a custom node without restoring original input?

### DISCUSSION THREAD

**Discussion by rfgamaral** [ueberdosis/tiptap#2049] [General Questions] [answered]
Here's a video demo of a custom node I built ([CodeSandbox](https://codesandbox.io/s/tiptap-h56o9) for testing):

https://user-images.githubusercontent.com/96476/137324598-8a8ed616-ee67-4de9-aa51-c5b2e8c8b044.mp4

You can see that pressing <kbd>Backspace</kbd> deletes the custom node, but restores the original input that triggered the creation of the custom node. How can I code the <kbd>Backspace</kbd> handler to delete the node **without** restoring `:D ` (using the example in the demo)? I'm having a hard time figuring this one out.

**Comment by philippkuehn** [comment]
I think you are using input rules, right? This is caused by the `undoInputRule` command on backspace. 

https://github.com/ueberdosis/tiptap/blob/main/packages/core/src/extensions/keymap.ts

You can add a custom backspace handler without using `undoInputRule`.

**Reply by rfgamaral** [reply]
@philippkuehn 

> I think you are using input rules, right? This is caused by the `undoInputRule` command on backspace.

Yea, I'm using `nodeInputRule`. I was just updating the OP with a link to the CodeSandbox while you wrote this.

> [`main`/packages/core/src/extensions/keymap.ts](https://github.com/ueberdosis/tiptap/blob/main/packages/core/src/extensions/keymap.ts)
> 
> You can add a custom backspace handler without using `undoInputRule`.

Just tried that, and unless I'm doing it wrong, it doesn't seem to work (see [emoji-replacer.ts](https://codesandbox.io/s/tiptap-emoji-replacer-bw4qo?file=/src/extensions/emoji-replacer.ts) on CodeSandbox).

**Reply by philippkuehn** [reply]
Oh you are right. This isn’t working because `first()` return `false`. After that, the default backspace handler is getting called which includes `undoInputRule`. Only if this backspace handler returns `false` too, the browser removes the content before. 

Unfortunately I do not have an idea right now to prevent that.

**Reply by rfgamaral** [reply]
But it should be possible with a custom Backspace handler, right? I've tried a few things (there's some code commented in the CodeSandbox), but I wasn't able to achieve what I want.

**Comment by rfgamaral** [comment, marked as answer]
I solved my issue with a custom `Backspace` handler:

```ts
    addKeyboardShortcuts() {
        return {
            Backspace: () =>
                this.editor.commands.command(({ tr, state }) => {
                    const { empty, anchor } = state.selection

                    if (!empty) {
                        return false
                    }

                    let isBackspaceHandled = false

                    state.doc.nodesBetween(anchor - 1, anchor, (node, pos) => {
                        if (node.type.name === 'emojiReplacer') {
                            tr.deleteRange(pos, pos + node.nodeSize)
                            isBackspaceHandled = true
                            return false
                        }
                    })

                    return isBackspaceHandled
                }),
        }
    },
```

It may not be the safest implementation for this [`emoji-replacer.ts`](https://codesandbox.io/s/tiptap-h56o9?file=/src/extensions/emoji-replacer.ts), but it seems to be working okay for now.
